datafusion/physical-plan/src/coalesce_batches.rs

```diff
@@ -259,45 +261,72 @@ impl Stream for CoalesceBatchesStream {
     }
 }
 
+/// Enumeration of possible states for `CoalesceBatchesStream`.
+/// It represents different stages in the lifecycle of a stream of record batches.
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum CoalesceBatchesStreamState {
+    // State to pull a new batch from the input stream.
+    Pull,
+    // State to return a buffered batch.
+    ReturnBuffer,
+    // State indicating that the stream is exhausted.
+    Exhausted,
+}
+
 impl CoalesceBatchesStream {
     fn poll_next_inner(
         self: &mut Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Option<Result<RecordBatch>>> {
-        // Get a clone (uses same underlying atomic) as self gets borrowed below
         let cloned_time = self.baseline_metrics.elapsed_compute().clone();
-
-        if self.is_closed {
-            return Poll::Ready(None);
-        }
         loop {
-            let input_batch = self.input.poll_next_unpin(cx);
-            // records time on drop
-            let _timer = cloned_time.timer();
-            match ready!(input_batch) {
-                Some(result) => {
-                    let Ok(input_batch) = result else {
-                        return Poll::Ready(Some(result)); // pass back error
+            match self.inner_state.clone() {
+                CoalesceBatchesStreamState::Pull => {
+                    // Attempt to pull the next batch from the input stream.
+                    let input_batch = match self.input.poll_next_unpin(cx) {
+                        Poll::Ready(t) => t,
+                        Poll::Pending => {
+                            // If the batch is not ready, do not alter the current state and return `Poll::Pending`.
+                            return Poll::Pending;
+                        }
                     };
```

Comment:
You can use the ready! macro here to simplify the code:https://doc.rust-lang.org/stable/std/task/macro.ready.html

Something like
```suggestion
                    let input_batch = ready!(self.input.poll_next_unpin(cx));
```